### PR TITLE
Adding a config util for bootstrap config

### PR DESCRIPTION
**Pull request by alyssawilk**
Starting work for #1162

### CODE REVIEW

**Comment by alyssawilk** [comment]
@mattklein123 once this is checked in, I'd like to extend it with helpers to make it easy for developers to use it rather than extending json flat files as we add new integration tests.  The best way to make sure it has the right helpers is to change one or more of the current integration tests to use proto by default and I want to check in before I start doing that.  We could also branch the tests and run both ways but with travis limitations I think that's probably a bad choice.

File: test/config/utility.cc [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+#include "test/test_common/environment.h"
+#include "test/test_common/network_utility.h"
+
+#include "api/filter/http_connection_manager.pb.h"
+
+namespace Envoy {
+
+std::string basic_config = R"EOF(
+{
+  "static_resources": {
```
Comment (mattklein123) [inline test/config/utility.cc]: Thoughts on using YAML for this so we can put comments directly into the config snippets that we will end up having? Not sure if worth it or not. (No real preference).

Comment (alyssawilk) [reply test/config/utility.cc]: Let's see what the config snippets end up being and then I can always change over if we think it's worth it, SG?

**Comment by alyssawilk** [comment]
Yeah, I'm hoping for most tests there will be things like
"addRoute("/long/url", "cluster_1")" or addRouteCorsPolicy(*route,
cors_string) (where the string can be pretty printed) and bundled with the
tests it should be easy to see what config is changed, and comments
explaining what is being tested.

We'll see how it plays out :-)

File: test/config/utility.cc (hunk 1 above)
Comment (alyssawilk) [reply test/config/utility.cc]: actually, @htuch want to weigh in here as well?

Comment (htuch) [reply test/config/utility.cc]: If humans are to consume/maintain, YAML is always best IMHO; JSON is just so much more verbose and noisy.

File: test/config/utility.cc [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    for (int j = 0; j < cluster->hosts_size(); ++j) {
+      auto* host_socket_addr = cluster->mutable_hosts(j)->mutable_socket_address();
+      ASSERT(ports.size() >= port_idx);
+      host_socket_addr->set_port_value(ports[port_idx++]);
+    }
+  }
+  ASSERT(port_idx == ports.size());
+}
+
+void ConfigHelper::setSourceAddress(std::string address_string) {
```
Comment (htuch) [inline test/config/utility.cc]: Nit: const std::string&

File: test/config/utility.cc [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+}
+
+void ConfigHelper::setUpstreamPorts(std::vector<uint32_t> ports) {
+  uint32_t port_idx = 0;
+  auto* static_resources = bootstrap_.mutable_static_resources();
+  for (int i = 0; i < bootstrap_.mutable_static_resources()->clusters_size(); ++i) {
+    auto* cluster = static_resources->mutable_clusters(i);
+    for (int j = 0; j < cluster->hosts_size(); ++j) {
+      auto* host_socket_addr = cluster->mutable_hosts(j)->mutable_socket_address();
+      ASSERT(ports.size() >= port_idx);
```
Comment (htuch) [inline test/config/utility.cc]: Do we just want to error out here instead of silently failing to set ports? Could be easier to debug if it fails fast.

File: test/config/utility.cc (hunk 1 above)
Comment (alyssawilk) [reply test/config/utility.cc]: Alas for my laziness, YAML it is then! =P
Will upload a new version, likely Monday.

File: test/config/utility.cc [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+#include "api/filter/http_connection_manager.pb.h"
+
+namespace Envoy {
+
+std::string basic_config = R"EOF(
+{
+  "static_resources": {
+    "listeners":[{
+      "name":"listener_0",
```
Comment (htuch) [inline test/config/utility.cc]: Why don't we have an `addListener()` that builds this? That way we can easily add multiple listeners.

File: test/config/utility.h [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+class ConfigHelper {
+public:
+  // Set up basic config, using the specified IpVersion for all connections: listeners, upstream,
+  // and admin connections.
+  ConfigHelper(const Network::Address::IpVersion version);
+
+  // Set the upstream ports.  The size of this vector must match the number of socket addresses
+  // across all configured clusters.
+  void setUpstreamPorts(std::vector<uint32_t> ports);
```
Comment (htuch) [inline test/config/utility.h]: Nit: const std::vector<>&

File: test/config/utility.cc (hunk 3 above)
Comment (alyssawilk) [reply test/config/utility.cc]: Yeah, my thought was that a fast fail was better.

File: test/config/utility.cc (hunk 4 above)
Comment (alyssawilk) [reply test/config/utility.cc]: I'd like that, but this TODO came from Matt who preferred the basic config in one place.

I figure we'll start with this, add the utilities as tests tweak config, and then maybe get used to just using the helpers?  Or we'll leave this here as a sample config.

File: test/config/utility.cc (hunk 3 above)
Comment (htuch) [reply test/config/utility.cc]: Switch to `RELEASE_ASSERT` then?

File: test/config/utility.cc (hunk 4 above)
Comment (htuch) [reply test/config/utility.cc]: SGTM, don't want to boil the ocean ;)

Comment (mattklein123) [reply test/config/utility.cc]: `addListener()` is OK with me, presumably with the listener template being defined somewhere. Really all I'm trying to maintain is some level of readability of what is going on in each test. As long as we have the right helpers and the right snippets to look back on it's probably fine.

File: test/config/utility.cc (hunk 3 above)
Comment (alyssawilk) [reply test/config/utility.cc]: Bah, forgot Envoy's ASSERT compiles out, thanks for the reminder

File: test/config/utility.h [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+  ConfigHelper(const Network::Address::IpVersion version);
+
+  // Set the upstream ports.  The size of this vector must match the number of socket addresses
+  // across all configured clusters.
+  void setUpstreamPorts(const std::vector<uint32_t>& ports);
+
+  // Set source_address in the bootstrap bind config.
+  void setSourceAddress(const std::string& address_string);
+
+  // Return the bootstrap configuration for dyamic edits.
```
Comment (htuch) [inline test/config/utility.h]: Quick question on API here. I kind of like immutability for config as a way of taming complexity. Could we do something similar to GFE, where we allow tests to provide a functional transform of a base config, rather than take a mutable reference to a global config, which might then be subject to arbitrary transform by multiple sites?

File: test/config/utility.cc [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+#include "common/protobuf/utility.h"
+
+#include "test/test_common/environment.h"
+#include "test/test_common/network_utility.h"
+
+#include "api/filter/http_connection_manager.pb.h"
+
+namespace Envoy {
+
+std::string basic_config = R"EOF(
```
Comment (mattklein123) [inline test/config/utility.cc]: const?

File: test/config/utility.cc [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+    name: listener_0
+    address:
+      socket_address:
+        address: 127.0.0.1
+        port_value: 0
+    filter_chains:
+      filters:
+        name: http_connection_manager
+        config:
+          httpFilters:
```
Comment (mattklein123) [inline test/config/utility.cc]: nit: Can we make all of these consistently snake_case?

File: test/config/utility.cc [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+            name: router
+            config:
+              deprecated_v1: true
```
Comment (mattklein123) [inline test/config/utility.cc]: I haven't been fully following but why is this necessary?

Comment (htuch) [reply test/config/utility.cc]: We just need to add createFilterFactoryFromProto to https://github.com/lyft/envoy/blob/master/source/server/config/http/router.h.

Comment (mattklein123) [reply test/config/utility.cc]: I see. Shouldn't we do that anyway for all of the internal filters? Or at least the ones we are using in v2 integration tests?

Comment (mattklein123) [reply test/config/utility.cc]: Or is the plan to have v2 filter config still use v1 JSON in existing cases?

Comment (htuch) [reply test/config/utility.cc]: Added a tracking issue at https://github.com/lyft/envoy/issues/1619. For router, it's trivial, for others, e.g. Mongo, it will involve a bit of work. Definitely something to mop up.

File: test/config/utility.h (hunk 6 above)
Comment (alyssawilk) [reply test/config/utility.h]: I'm not convinced arbitrary edits at multiple sites are more or less clear than closures added at multiple sites.  Still given no one is using it it's easy enough to lock it down for now.

File: test/config/utility.cc (hunk 8 above)
Comment (alyssawilk) [reply test/config/utility.cc]: Apparently we can.  Fascinating.
(was just editing what proto.DebugString() spat out)

File: test/config/utility.cc [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+  access_log_path: /dev/null
+  address:
+    socket_address:
+      address: 127.0.0.1
+      port_value: 0
+)EOF";
+
+ConfigHelper::ConfigHelper(const Network::Address::IpVersion version) {
+  // Add the default listener.
+  addListener("listener_0", version);
```
Comment (mattklein123) [inline test/config/utility.cc]: I'm a little confused what this is doing. Don't we want this to specify a filter stack somehow and/or some other listener config template?

Comment (alyssawilk) [reply test/config/utility.cc]: Oh, I left it pulling the filter stack from the basic config.

How about I roll this bit back, and we can refine and add helpers as we port tests over.
